packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_withdraw_from_vault.test.ts

```diff
@@ -0,0 +1,372 @@
+    })
+  })
+
+  describe('fail cases', () => {
+    it('should fail with invalid vaultId', async () => {
+      const destination = await tGroup.get(0).generateAddress()
+      const decoded = fromAddress(destination, 'regtest')
+      const script = decoded?.script as Script
+
+      const txn = await builder.loans.withdrawFromVault({
+        vaultId: '0'.repeat(64),
+        to: script,
+        tokenAmount: { token: 1, amount: new BigNumber(1) }
+      }, script)
+
+      const promise = sendTransaction(tGroup.get(0).container, txn)
+      await expect(promise).rejects.toThrow(DeFiDRpcError)
+      await expect(promise).rejects.toThrow('')
+    })
+
+    it('should failed as insufficient collateral token', async () => {
+      const script = await providers.elliptic.script()
+      const txn = await builder.loans.withdrawFromVault({
+        vaultId: vaultId1,
+        to: script,
+        tokenAmount: { token: 0, amount: new BigNumber(99999) }
+      }, vaultOwnerScript)
+
+      const promise = sendTransaction(tGroup.get(0).container, txn)
+      await expect(promise).rejects.toThrow(DeFiDRpcError)
+      await expect(promise).rejects.toThrow('')
+    })
+
+    it('should failed without valid auth (not my vault)', async () => {
+      const anotherMnProviders = await getProviders(tGroup.get(1).container)
+      anotherMnProviders.setEllipticPair(WIF.asEllipticPair(GenesisKeys[1].owner.privKey))
+      await fundEllipticPair(tGroup.get(1).container, anotherMnProviders.ellipticPair, 10)
+
+      // MN 2 sign a withdrawal and withdraw from MN1's vault
+      const anotherMnBuilder = new P2WPKHTransactionBuilder(anotherMnProviders.fee, anotherMnProviders.prevout, anotherMnProviders.elliptic, RegTest)
+      const destination = await tGroup.get(1).generateAddress() // another MN's address
+      const decoded = fromAddress(destination, 'regtest')
+      const script = decoded?.script as Script
+
+      const txn = await anotherMnBuilder.loans.withdrawFromVault({
+        vaultId: vaultId1, // belongs to MN 1
+        to: script,
+        tokenAmount: { token: 0, amount: new BigNumber(1) }
+      }, vaultOwnerScript)
+
+      const promise = sendTransaction(tGroup.get(1).container, txn)
+      await expect(promise).rejects.toThrow('tx must have at least one input from token owner')
+    })
+
+    it('should failed as vault must contain min 50% of DFI', async () => {
+      const destination = await tGroup.get(0).generateAddress()
+      const decoded = fromAddress(destination, 'regtest')
+      const script = decoded?.script as Script
+
+      // set oracle price above spent the utxos and change go into b58 address
+      // causing no prevout at address derived directly from elliptic pair, fund for utxo again
+      await fundEllipticPair(tGroup.get(0).container, providers.ellipticPair, 10)
+
+      const what = await tGroup.get(0).rpc.loan.getVault(dualCollateralVault)
+      console.log(what)
```

Comment:
remove what?😅

Reply:
oops, the bug was on defid, this test wasn't able to pass. Forgotten them after too long